node/src/components/linear_chain/state.rs

```diff
@@ -223,33 +240,68 @@ impl LinearChain {
         block: Box<Block>,
         execution_results: HashMap<DeployHash, ExecutionResult>,
     ) -> Outcomes {
-        let mut outcomes = vec![];
+        vec![Outcome::StoreBlock(block, execution_results)]
+    }
+
+    pub(super) fn handle_put_block(&mut self, block: Box<Block>) -> Outcomes {
+        let mut outcomes = Vec::new();
         let signatures = self.new_block(&*block);
+        self.latest_block = Some(*block.clone());
+        if let Some(key_block_info) = KeyBlockInfo::maybe_from_block_header(block.header()) {
+            let current_era = key_block_info.era_id();
+            self.key_block_info.insert(current_era, key_block_info);
+            if let Some(old_era_id) = self.lowest_acceptable_era_id(current_era).checked_sub(1) {
+                self.key_block_info.remove(&old_era_id);
+            }
+        }
         if !signatures.is_empty() {
             let mut block_signatures = BlockSignatures::new(*block.hash(), block.header().era_id());
             for sig in signatures.iter() {
                 block_signatures.insert_proof(sig.public_key(), sig.signature());
             }
-            outcomes.push(Outcome::StoreBlockSignatures(block_signatures));
+            let should_upgrade = self.should_upgrade(&block_signatures);
+            outcomes.push(Outcome::StoreBlockSignatures(
+                block_signatures,
+                should_upgrade,
+            ));
             for signature in signatures {
                 if signature.is_local() {
                     outcomes.push(Outcome::Gossip(Box::new(signature.to_inner().clone())));
                 }
                 outcomes.push(Outcome::AnnounceSignature(signature.take()));
             }
         };
-        let block_hash = *block.hash();
-        outcomes.push(Outcome::StoreBlock(block));
-        outcomes.push(Outcome::StoreExecutionResults(
-            block_hash,
-            execution_results,
-        ));
+        outcomes.push(Outcome::AnnounceBlock(block));
         outcomes
     }
 
-    pub(super) fn handle_put_block(&mut self, block: Box<Block>) -> Outcomes {
-        self.set_latest_block(*block.clone());
-        vec![Outcome::AnnounceBlock(block)]
+    fn should_upgrade(&self, signatures: &BlockSignatures) -> bool {
+        let signed_kb_info = match self
+            .key_block_info
+            .get(&signatures.era_id.saturating_add(1))
```

Comment:
Why do we have to add 1 to it? When we store key block it's stored under that block's era_id which isn't incremented for key block.

Reply:
Added `checked_add` in https://github.com/casper-network/casper-node/pull/2341/commits/3f6b042dd560cd8d6cd73613f9acbf6d73f7e789.